src/Build/Logging/FancyLogger/FancyLoggerNode.cs

```diff
@@ -4,148 +4,136 @@
-                    Id = $"message-{id}";
-                    break;
-                case FancyLoggerNodeType.Warning:
-                    Id = $"warning-{id}";
-                    break;
-                case FancyLoggerNodeType.Error:
-                    Id = $"error-{id}";
-                    break;
-                default:
-                    Id = id.ToString(); break;
+                FancyLoggerBuffer.DeleteLine(CurrentTargetLine.Id);
+                return;
             }
+            string currentTargetLineContents = $"   └── {CurrentTargetNode.TargetName} : {CurrentTargetNode.CurrentTaskNode?.TaskName ?? String.Empty}";
+            if (CurrentTargetLine == null) CurrentTargetLine = FancyLoggerBuffer.WriteNewLineAfter(currentTargetLineContents, Line.Id);
+            else FancyLoggerBuffer.UpdateLine(CurrentTargetLine.Id, currentTargetLineContents);
         }
-        public FancyLoggerNode? Find(string id)
+
+        public void AddTarget(TargetStartedEventArgs args)
+        {
+            CurrentTargetNode = new FancyLoggerTargetNode(args);
+        }
+        public void AddTask(TaskStartedEventArgs args)
         {
-            // If self
-            if(Id == id) return this;
-            // If no children
-            if(Children.Count == 0) return null;
-            // Iterate
-            foreach (var child in Children)
+            // Get target id
+            int targetId = args.BuildEventContext!.TargetId;
+            if (CurrentTargetNode?.Id == targetId)
             {
-                FancyLoggerNode? node = child.Value.Find(id);
-                if (node != null) return node;
+                CurrentTargetNode.AddTask(args);
             }
-            return null;
         }
+        public void AddMessage(BuildMessageEventArgs args)
+        {
+            if (args.Importance != MessageImportance.High) return;
+            AdditionalDetails.Add(new FancyLoggerMessageNode(args));
+        }
+        public void AddWarning(BuildWarningEventArgs args)
+        {
+            AdditionalDetails.Add(new FancyLoggerWarningNode(args));
+        }
+        public void AddError(BuildErrorEventArgs args)
+        {
+            AdditionalDetails.Add(new FancyLoggerErrorNode(args));
+        }
+    }
 
-        public void Add(FancyLoggerNode node)
+    public class FancyLoggerTargetNode
+    {
+        public int Id;
+        public string TargetName;
+        public FancyLoggerTaskNode? CurrentTaskNode;
+        public FancyLoggerTargetNode(TargetStartedEventArgs args)
+        {
+            Id = args.BuildEventContext!.TargetId;
+            TargetName = args.TargetName;
+        }
+        public void AddTask(TaskStartedEventArgs args)
         {
-            if (Children.ContainsKey(node.Id)) return;
-            Children.Add(node.Id, node);
-            node.Depth = Depth + 1;
-            node.Parent = this;
+            CurrentTaskNode = new FancyLoggerTaskNode(args);
```

Comment:
Similar to targets, the assumption that one task executes to completion before the next task executes may not be correct. With IBuildEngine3, we added "Yield" and "Reacquire" calls that a task can do to let something else execute while the task does something in the background that MSBuild doesn't have to care about. That may mean you'd have multiple tasks executing "simultaneously"...

Reply:
`Yield` yields the whole project, so there shouldn't ever be multiple tasks in flight for a single project. This should be ok for the current per-project display scheme.